server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -320,4 +346,298 @@ public DirectoryReader getDirectoryReader() {
         assert reader instanceof DirectoryReader : "expected an instance of DirectoryReader, got " + reader.getClass();
         return (DirectoryReader) reader;
     }
+
+    /**
+     * iFace which implements the query timeout / cancellation logic
+     */
+    public interface Cancellable {
+
+        boolean isEnabled();
+        void checkCancelled();
+        default void checkDirReaderCancelled() {
```

Comment:
These two functions should not be part of the interface, checkCancelled should be enough

Reply:
I don't like that much either, I exposed the checkDirReaderCancelled() so that I we can have tests that differentiate between exit in the searcher and exit from the dirReader: https://github.com/elastic/elasticsearch/pull/52822/files#diff-2abce48f3a52657ce3740afade6d5f8fR132
But as discussed I could just unit test the Cancellable reader instead.